iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -137,99 +154,440 @@ public static RegistryManager getRegistyManager(String iotHubConnectionString) t
+                    }
+                }
 
-            //x509 device to house the x509 module under test
-            String deviceId = "java-test-x509-device-".concat(UUID.randomUUID().toString());
-            Device device = Device.createDevice(deviceId, SELF_SIGNED);
-            device.setThumbprintFinal(x509Thumbprint, x509Thumbprint);
-            device = Tools.addDeviceWithRetry(getRegistyManager(iotHubConnectionString), device);
+                log.debug("Acquiring test module from testX509ModuleQueue");
+                testModuleIdentity = testX509ModuleQueue.remove();
+            }
 
-            //x509 module client under test
-            String moduleId = "java-test-x509-module-".concat(UUID.randomUUID().toString());
-            Module module = Module.createModule(deviceId, moduleId, SELF_SIGNED);
-            module.setThumbprintFinal(x509Thumbprint, x509Thumbprint);
-            module = Tools.addModuleWithRetry(getRegistyManager(iotHubConnectionString), module);
-            ModuleClient moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, device, module), protocol, sslContext);
+            SSLContext sslContext = SSLContextBuilder.buildSSLContext(IntegrationTest.x509CertificateGenerator.getPublicCertificate(), IntegrationTest.x509CertificateGenerator.getPrivateKey());
+            ModuleClient moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, testModuleIdentity.device, testModuleIdentity.module), protocol, sslContext);
+            testModuleIdentity.setModuleClient(moduleClient);
+            return testModuleIdentity;
+        }
+    }
 
-            return new TestModuleIdentity(moduleClient, device, module);
+    /**
+     * If the environment variable RECYCLE_TEST_IDENTITIES is set to true, then the identities will be recycled. Otherwise they
+     * will be removed from the IoT Hub's registry.
+     * @param testIdentities The test identities to dispose/recycle.
+     */
+    public static void disposeTestIdentities(Iterable<? extends TestIdentity> testIdentities, String iotHubConnectionString)
+    {
+        if (!IntegrationTest.recycleIdentities)
+        {
+            for (TestIdentity testIdentity : testIdentities)
+            {
+                disposeTestIdentity(testIdentity, iotHubConnectionString);
+            }
         }
         else
         {
-            throw new IllegalArgumentException("Test code has not been written for this path yet");
+            List<Device> devicesToDelete = new ArrayList<>();
+            for (TestIdentity testIdentity : testIdentities)
+            {
+                devicesToDelete.add(testIdentity.getDevice());
+            }
+
+            // No need to dispose of modules in particular.
+            // Deleting a device that has a module will delete both the device and module.
+            disposeTestDevices(devicesToDelete, iotHubConnectionString);
```

Comment:
I think I'm mis-reading something here. If RECYCLE_TEST_IDENTITIES is set to true, why are we still removing all the test devices?

Reply:
Oops, I had this backwards. Line 450 should have been "if recycleIdentities then dispose each identity" where the method call to dispose test identity just sticks each identity into the appropriate recycling queue.